### PR TITLE
Intercept events from media session

### PR DESCRIPTION
**Pull request by doughsay**
This fixes https://github.com/doublesymmetry/react-native-track-player/issues/1582

This allows us to act on events caused by other things controlling the media session, not just the notification buttons, e.g. triggering play/pause or fast forward/rewind from bluetooth headphones.

I tested this with the react-native-track-player example app (with modifications to pass on the new events) and it appears to work.

Please let me know if anything here is wrong and I will happily fix it.

### CODE REVIEW

File: kotlin-audio/src/main/java/com/doublesymmetry/kotlinaudio/notification/NotificationManager.kt [hunk 1]
```diff
@@ -162,6 +162,18 @@ class NotificationManager internal constructor(private val context: Context, pri
         }
 
     init {
+        mediaSession.setCallback(object : MediaSessionCompat.Callback() {
+            override fun onPause() {
+                super.onPause()
+                event.updateOnNotificationButtonTapped(NotificationButton.PAUSE(null))
```
[the remark this answers is not in the document]
Comment (doughsay) [reply kotlin-audio/src/main/java/com/doublesymmetry/kotlinaudio/notification/NotificationManager.kt:168]: So my first attempt was to add new `MediaSessionCallback.PLAY` and `MediaSessionCallback.PAUSE` and then in the handler do `event.updateOnMediaSessionCallbackTriggered(MediaSessionCallback.PAUSE(...)` etc., but all that succeeded in doing was apparently intercepting the bluetooth events, but not actually acting on them.  Let me clarify:

Current behavior in KotlinAudio 0.1.34: Bluetooth events DO play/pause audio, but the service does not receive the event.
Behavior with MediaSessionCallback handlers: Bluetooth events don't do anything. Audio keeps playing, no events are received.
Behavior with my current change: Bluetooth events play/pause audio AND fire events, but tapping notification buttons sends 2 events.

If the media session callbacks are the right way to do it, then my next question will be: how do I actually forward the event to the trackplayer service?  Is this handled in the trackplayer package maybe?

**Comment by doughsay** [comment]
> this is looking good! I have a couple of questions:

Thanks!  Here's what I know:

1. I was positive that the events weren't duplicated for some reason, but I double checked, and you're right...  if you tap the buttons in the notification the events double fire 😭 
2. I can do that if it makes sense, I was just trying to keep the changes separate to make it easier to see just the relevant changes.
3. I don't have any bluetooth devices that send any other events, but I assume there must be some way to simulate this in the emulator?  I'll look into that to see what I can find.

So, the duplicated events makes this a no-go anyway, so any pointers on how to proceed would be much appreciated.  Please see my other inline comment for some more details and questions.

File: kotlin-audio/src/main/java/com/doublesymmetry/kotlinaudio/notification/NotificationManager.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (doughsay) [reply kotlin-audio/src/main/java/com/doublesymmetry/kotlinaudio/notification/NotificationManager.kt:168]: I think I answered my own question.  I'd need to add stuff here: https://github.com/doublesymmetry/react-native-track-player/blob/main/android/src/main/java/com/doublesymmetry/trackplayer/service/MusicService.kt#L428

[the remark this answers is not in the document]
Comment (doughsay) [reply kotlin-audio/src/main/java/com/doublesymmetry/kotlinaudio/notification/NotificationManager.kt]: Ok, I switched to using media session callback events and updates track-player locally to forward them on to the service, and that fixes everything.  No more duplicate events!

**Comment by doughsay** [comment]
Updates:

1. The latest commit fixes duplicate events (but requires a change to track-player that I'll open as a separate PR once this is accepted)
2. Maybe we want to just combine the two PRs into a general "media session events" PR that handles all relevant events?
3. The emulator does not support bluetooth, the best we can do is emulate the keypress events to test functionality. I can do this, but don't really know which key events are common on bluetooth devices.  But maybe it's best to separate the concepts a bit and just more generally handle all media session events.

So which events would we be interested in? Here's my proposal:

* onFastForward
* onPause
* onPlay
* onRewind
* onSeekTo (from your PR)
* onSkipToNext
* onSkipToPrevious
* onStop

If we handle all these and forward them to the service in track-player then it will probably handle most use-cases.

**Comment by jspizziri** [comment]
@doughsay tbh, I think we're going to need all of them. [They were all implemented in the `v2.1` version of RNTP](https://github.com/doublesymmetry/react-native-track-player/blob/v2.1/android/src/main/java/com/guichaguri/trackplayer/service/metadata/ButtonEvents.java)

I think ultimately, we're going to need to remove the events that are manually emitted when a button is clicked in favor of the exoplayer ones so can cover all these bases.

@mpivchev at this point we're really going to need you to weigh in here.

**Comment by doughsay** [comment]
What you say seems reasonable, but I'm not really going to know how to implement the more complex ones. The simple no-argument ones I can do, but things like `onPlayFromSearch` or `onPlayFromMediaId` I don't even know how to test.

**Comment by jspizziri** [comment]
@doughsay just following up here. I don't feel comfortable moving forward without @mpivchev 's approval on the plan I outlined. 

I'd be fine leaving out some of the more complicated things like `onPlayFromSearch` etc in this PR and handling them separately elsewhere, everything else alone would be a big win.

@mpivchev would love your feedback here!

**Comment by doughsay** [comment]
Yeah, I understand. I'll rebase onto your PR to get seekTo and add the other trivial ones and maybe we can get approval to merge that for now.

**Comment by doughsay** [comment]
@jspizziri @mpivchev I think this is ready for review. We're still missing a few events that I'm not sure how to properly implement (e.g. `onPlayFromSearch`, `onPlayFromMediaId`), but the rest of them I think are good to go.  I've also opened a PR in RNTP to act on these events and forward them to the player service: https://github.com/doublesymmetry/react-native-track-player/pull/1593

**Comment by doughsay** [comment]
> @doughsay good work!
> 
> I'm assuming that we're getting the double-fired event on many of these correct?

Nope!  No duplicate events.  Not sure why, but everything in this PR + https://github.com/doublesymmetry/react-native-track-player/pull/1593 is working for 100% in my app.

You might still be right that we can remove the button events, because if they're not being intercepted, then the media session events will be, but right now it seems that handling the button taps manually causes it to not propagate to the media session.

**Comment by mpivchev** [comment]
Ultimately we want to rid ourselves of the forked ExoPlayer we use, and from what I gathered here, the plan is to not have `onMediaSessionCallbackTriggered` and `onNotificationButtonTapped` events and instead use the events provided by ExoPlayer, as they handle all cases?

**Comment by jspizziri** [comment]
@mpivchev yes, that's the case. I'll work with @doughsay to make this happen.

**Comment by mpivchev** [comment]
This is handled here #38

**Pull request closed without merging**